@here/harp-texture-atlas-generator/src/ImageFactory.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { RGBA } from "./ColorUtils";
+import { FileSystem, ImageFormat } from "./FileSystem";
+
+export type Color = RGBA;
+
+export interface ImageDecoder {
+    readonly width: number;
+    readonly height: number;
+
+    getPixelAt(x: number, y: number): Color;
+
+    copy(): ImageEncoder;
+}
+
+export interface ImageEncoder {
+    readonly width: number;
+    readonly height: number;
+
+    setPixelAt(x: number, y: number, color: Color): void;
+    getPixelAt(x: number, y: number): Color;
+
+    copy(): ImageEncoder;
+    resize(width: number, height: number): ImageEncoder;
+
+    write(filePath: string): Promise<any>;
+    writeSync(filePath: string): void;
+}
+
+// TODO: Figure out some better pattern - static vs instance interface
```

Comment:
As proposed in other comment.
In JS and other functional languages, `factories` can be "just functions.

In this case it would be:

```
export type ImageDecoderLoader/Factory/Builder/Constructor = (filePath: string) => Promise<ImageDecode>
```

Same in many other places.